### ISSUE TITLE
[Core][nightly-test] long_running_many_drivers is flaky due to segmentation fault in GcsRpcClient GcsSubscribePool

### ISSUE THREAD

**Issue by scv119** [ray-project/ray#30679]
### What happened + What you expected to happen

Segmentation fault in GcsRpcClient

https://console.anyscale-staging.com/o/anyscale-internal/projects/prj_qC3ZfndQWYYjx2cz8KWGNUL4/clusters/ses_du3n1czxzsyx9pk1dt34e96h

```
run_driver pid=111242) *** SIGSEGV received at time=1669418940 on cpu 4 ***
(run_driver pid=111242) PC: @     0x7f77fc7ec955  (unknown)  ray::rpc::GcsRpcClient::GcsSubscriberPoll()::{lambda()#3}::operator()()
(run_driver pid=111242)     @     0x7f77fda9f420       3600  (unknown)
(run_driver pid=111242)     @     0x7f77fc7ed22a        560  ray::rpc::GcsRpcClient::GcsSubscriberPoll()
(run_driver pid=111242)     @     0x7f77fc7ed50b        176  ray::gcs::(anonymous namespace)::GcsSubscriberClient::PubsubLongPolling()
(run_driver pid=111242)     @     0x7f77fc851be1        272  ray::pubsub::Subscriber::MakeLongPollingPubsubConnection()
(run_driver pid=111242)     @     0x7f77fc85640b        528  ray::pubsub::Subscriber::HandleLongPollingResponse()
(run_driver pid=111242)     @     0x7f77fc856812         80  std::_Function_handler<>::_M_invoke()
(run_driver pid=111242)     @     0x7f77fc7ed763        128  std::_Function_handler<>::_M_invoke()
(run_driver pid=111242)     @     0x7f77fc7edfef        128  ray::rpc::GcsRpcClient::GcsSubscriberPoll()::{lambda()#2}::operator()()
(run_driver pid=111242)     @     0x7f77fc7e9bd1         64  ray::rpc::ClientCallImpl<>::OnReplyReceived()
(run_driver pid=111242)     @     0x7f77fc638235         32  std::_Function_handler<>::_M_invoke()
(run_driver pid=111242)     @     0x7f77fc95d3a6         80  EventTracker::RecordExecution()
(run_driver pid=111242)     @     0x7f77fc90007e         48  std::_Function_handler<>::_M_invoke()
(run_driver pid=111242)     @     0x7f77fc9001f6        112  boost::asio::detail::completion_handler<>::do_complete()
(run_driver pid=111242)     @     0x7f77fcd9913b        128  boost::asio::detail::scheduler::do_run_one()
(run_driver pid=111242)     @     0x7f77fcd9a371        192  boost::asio::detail::scheduler::run()
(run_driver pid=111242)     @     0x7f77fcd9a5e0         64  boost::asio::io_context::run()
(run_driver pid=111242)     @     0x7f77fc6c05bd        240  ray::core::CoreWorker::RunIOService()
(run_driver pid=111242)     @     0x7f77fcece390  (unknown)  execute_native_thread_routine
(run_driver pid=111242)     @     0x556977c2b340  (unknown)  (unknown)
(run_driver pid=111242)     @     0x7f77fc6180a0  (unknown)  (unknown)
(run_driver pid=111242)     @ 0x90008b6311e90789  (unknown)  (unknown)
(run_driver pid=111242) [2022-11-25 15:29:00,586 E 414925 415123] logging.cc:361: *** SIGSEGV received at time=1669418940 on cpu 4 ***
(run_driver pid=111242) [2022-11-25 15:29:00,586 E 414925 415123] logging.cc:361: PC: @     0x7f77fc7ec955  (unknown)  ray::rpc::GcsRpcClient::GcsSubscriberPoll()::{lambda()#3}::operator()()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fda9f420       3600  (unknown)
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc7ed22a        560  ray::rpc::GcsRpcClient::GcsSubscriberPoll()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc7ed50b        176  ray::gcs::(anonymous namespace)::GcsSubscriberClient::PubsubLongPolling()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc851be1        272  ray::pubsub::Subscriber::MakeLongPollingPubsubConnection()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc85640b        528  ray::pubsub::Subscriber::HandleLongPollingResponse()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc856812         80  std::_Function_handler<>::_M_invoke()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc7ed763        128  std::_Function_handler<>::_M_invoke()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc7edfef        128  ray::rpc::GcsRpcClient::GcsSubscriberPoll()::{lambda()#2}::operator()()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc7e9bd1         64  ray::rpc::ClientCallImpl<>::OnReplyReceived()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc638235         32  std::_Function_handler<>::_M_invoke()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc95d3a6         80  EventTracker::RecordExecution()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc90007e         48  std::_Function_handler<>::_M_invoke()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc9001f6        112  boost::asio::detail::completion_handler<>::do_complete()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fcd9913b        128  boost::asio::detail::scheduler::do_run_one()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fcd9a371        192  boost::asio::detail::scheduler::run()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fcd9a5e0         64  boost::asio::io_context::run()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fc6c05bd        240  ray::core::CoreWorker::RunIOService()
(run_driver pid=111242) [2022-11-25 15:29:00,593 E 414925 415123] logging.cc:361:     @     0x7f77fcece390  (unknown)  execute_native_thread_routine
(run_driver pid=111242) [2022-11-25 15:29:00,596 E 414925 415123] logging.cc:361:     @     0x556977c2b340  (unknown)  (unknown)
(run_driver pid=111242) [2022-11-25 15:29:00,607 E 414925 415123] logging.cc:361:     @     0x7f77fc6180a0  (unknown)  (unknown)
(run_driver pid=111242) [2022-11-25 15:29:00,609 E 414925 415123] logging.cc:361:     @ 0x90008b6311e90789  (unknown)  (unknown)Traceback (most recent call last):
  File "workloads/many_drivers.py", line 127, in <module>
    ray.get(ready_id)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/worker.py", line 2309, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(CalledProcessError): ray::run_driver() (pid=111242, ip=172.31.193.87)
  File "workloads/many_drivers.py", line 95, in run_driver
    output = run_string_as_driver(driver_script, encode="utf-8")
subprocess.CalledProcessError: Command '['/home/ray/anaconda3/bin/python', '-']' died with <Signals.SIGSEGV: 11>.
2022-11-25 15:29:03,724 WARNING worker.py:1851 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: f72ceba78fbe7094e3af0e3358d73019dd645eae01000000 Worker ID: a81a0df74b0332d920755d92e2c86eb49a9d1c5d7b27e624ebd66c1d Node ID: 7dab513dfe8546e6c325092e1e47670f65f75c089b5a3644cb3fb278 Worker IP address: 172.31.193.87 Worker port: 46153 Worker PID: 201153 Worker exit type: SYSTEM_ERROR Worker exit detail: Worker unexpectedly exits with a connection error code 2. End of file. There are some potential root causes. (1) The process is killed by SIGKILL by OOM killer due to high memory usage. (2) ray stop --force is called. (3) The worker is crashed unexpectedly due to SIGSEGV or other unexpected errors.

(run_driver pid=111242) Fatal Python error: Segmentation fault
```

### Versions / Dependencies

lastest master

### Reproduction script

long_running_many_drivers

### Issue Severity

High: It blocks me from completing my task.

**Comment by rkooo567** [comment]
Guess this PR is the most suspicious? https://github.com/ray-project/ray/pull/30131

**Comment by stephanie-wang** [comment]
@iycheng I tried rerunning the failed test and it passed, so I think it is flaky. But I took another look at the PR Sang linked and I think it is most likely the culprit. This [access](https://github.com/iycheng/ray/blob/6d47b9b12c9328fb156cfa96249f21c705f6fac8/src/ray/rpc/server_call.h#L252) might be the issue.

**Comment by iycheng** [comment]
Yeah, OK to revert it.

**Comment by stephanie-wang** [comment]
Closed by #30737.
